File: IAM/tools/replication/ldap_sdiff.go
package main

import (
	"database/sql"
	"flag"
	"fmt"
	_ "github.com/ibmdb/go_ibm_db"
	"os"
	"strings"
)

// Type for entry information
type ldapEntry struct {
	dn_trunc         string
	dn               string
	modify_timestamp string
}

var verbose = 0

func listAllEntries(DBconn *sql.DB, schema string, out chan<- ldapEntry) {
	defer close(out)
	listAllEntries := []string{
		"select dn_trunc, modify_timestamp ",
		"from %s.ldap_entry order by dn_trunc "}
	listAllEntriesSQLTemplate := strings.Join(listAllEntries, "")
	listAllEntriesSQL := fmt.Sprintf(listAllEntriesSQLTemplate, schema)
	statement, err := DBconn.Prepare(listAllEntriesSQL)
	if err != nil {
		fmt.Println("Error on Prepare: ", err.Error())
		return
	}
	rows, err := statement.Query()
	if err != nil {
		fmt.Println("Error on Query: ", err.Error())
		return
	}
	defer rows.Close()

	for rows.Next() {
		var dn_trunc, modify_timestamp string
		err = rows.Scan(&dn_trunc, &modify_timestamp)
		if err != nil {
			fmt.Println("Error on Scan: ", err.Error())
			return
		}
		out <- ldapEntry{dn_trunc, dn_trunc, modify_timestamp}
	}
	return
}

func compareAllEntryModifyTimestamps(firstDB *sql.DB, secondDB *sql.DB, schema1 string, schema2 string) error {
	fmt.Println("Reporting dn_trunc and modify_timestamp for any conflicting entries")
	fmt.Println("-------------------------------------------------------------------")

	ldap1Entries := make(chan ldapEntry)
	ldap2Entries := make(chan ldapEntry)

	go listAllEntries(firstDB, schema1, ldap1Entries)
	go listAllEntries(secondDB, schema2, ldap2Entries)

	ldap1Entry := <-ldap1Entries
	ldap2Entry := <-ldap2Entries

	for ldap1Entry.dn_trunc != "" && ldap2Entry.dn_trunc != "" {
		if verbose > 1 {
			fmt.Println("ldap1Entry: ", ldap1Entry.dn_trunc)
			fmt.Println("ldap2Entry: ", ldap2Entry.dn_trunc)
		}
		switch {
		case ldap1Entry.dn_trunc == ldap2Entry.dn_trunc:
			if ldap1Entry.modify_timestamp != ldap2Entry.modify_timestamp {
				fmt.Printf("Mismatching timestamps for %s: %s != %s\n", ldap1Entry.dn_trunc, ldap1Entry.modify_timestamp, ldap2Entry.modify_timestamp)
			}
			ldap1Entry = <-ldap1Entries
			ldap2Entry = <-ldap2Entries

		case ldap1Entry.dn_trunc < ldap2Entry.dn_trunc:
			fmt.Printf("Missing entry on second server: %s\n", ldap1Entry.dn_trunc)
			ldap1Entry = <-ldap1Entries

		case ldap1Entry.dn_trunc > ldap2Entry.dn_trunc:
			fmt.Printf("Missing entry on first server: %s\n", ldap2Entry.dn_trunc)
			ldap2Entry = <-ldap2Entries

		}
	}
	for ldap1Entry.dn_trunc != "" {
		fmt.Printf("Missing entry on second server: %s\n", ldap1Entry.dn_trunc)
		ldap1Entry = <-ldap1Entries
	}

	for ldap2Entry.dn_trunc != "" {
		fmt.Printf("Missing entry on first server: %s\n", ldap2Entry.dn_trunc)
		ldap2Entry = <-ldap2Entries
	}

	return nil
}

func CreateConn(con string) *sql.DB {
	db, err := sql.Open("go_ibm_db", con)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	return db
}

func DoUsage(message string) {
	fmt.Println(strings.TrimSpace(`
usage: ldap_sdiff.go [-h] --dbname1 DBNAME [--hostname1 HOSTNAME]
                       [--port1 PORT] [--schema1 SCHEMA] [--userid1 USERID] --password1 PASSWORD
                       --dbname2 DBNAME [--hostname2 HOSTNAME]
                       [--port2 PORT] [--schema2 SCHEMA] [--userid2 USERID] --password2 PASSWORD
`))
	if message != "" {
		fmt.Println(message)
	}
	os.Exit(1)
}

func DoHelp() {
	fmt.Println(strings.TrimSpace(`
usage: ldap_sdiff.go [-h] --dbname1 DBNAME [--hostname1 HOSTNAME]
                       [--port1 PORT] [--schema1 SCHEMA] [--userid1 USERID] --password1 PASSWORD
                       --dbname2 DBNAME [--hostname2 HOSTNAME]
                       [--port2 PORT] [--schema2 SCHEMA] [--userid2 USERID] --password2 PASSWORD
Provide DB2 connection details to determine replication status.

optional arguments:
  -h, --help           show this help message and exit
  --dbname1 DBNAME      DB2 Database Name underlying LDAP.
  --hostname1 HOSTNAME  Hostname of LDAP server (defaults to localhost).
  --port1 PORT          Port# DB2 is listening on (defaults to 50000).
  --schema1 SCHEMA      DB2 Table name schema (defaults to userid).
  --userid1 USERID      Userid to connect to DB2 (defaults to dbname).
  --password1 PASSWORD  Password to connect to DB2.
  --dbname2 DBNAME      DB2 Database Name underlying LDAP.
  --hostname2 HOSTNAME  Hostname of LDAP server (defaults to localhost).
  --port2 PORT          Port# DB2 is listening on (defaults to 50000).
  --schema2 SCHEMA      DB2 Table name schema (defaults to userid).
  --userid2 USERID      Userid to connect to DB2 (defaults to dbname).
  --password2 PASSWORD  Password to connect to DB2.
`))
	os.Exit(1)
}

// Set to different things by who is building
var Version = "sandbox"

func main() {
	fs := flag.NewFlagSet("ldap_sdiff", flag.ContinueOnError)
	dbname1Arg := fs.String("dbname1", "", "DB2 Database Name underlying LDAP.")
	hostname1Arg := fs.String("hostname1", "localhost", "Hostname of LDAP server (defaults to localhost).")
	port1Arg := fs.Int("port1", 50000, "Port# DB2 is listening on (defaults to 50000).")
	schema1Arg := fs.String("schema1", "", "DB2 Table name schema (defaults to userid).")
	userid1Arg := fs.String("userid1", "", "Userid to connect to DB2 (defaults to dbname).")
	password1Arg := fs.String("password1", "", "Password to connect to DB2..")
	dbname2Arg := fs.String("dbname2", "", "DB2 Database Name underlying LDAP.")
	hostname2Arg := fs.String("hostname2", "localhost", "Hostname of LDAP server (defaults to localhost).")
	port2Arg := fs.Int("port2", 50000, "Port# DB2 is listening on (defaults to 50000).")
	schema2Arg := fs.String("schema2", "", "DB2 Table name schema (defaults to userid).")
	userid2Arg := fs.String("userid2", "", "Userid to connect to DB2 (defaults to dbname).")
	password2Arg := fs.String("password2", "", "Password to connect to DB2..")
	verboseArg := fs.Int("verbose", 0, "Level of debugging (defaults to 0 - none).")
	help := fs.Bool("help", false, "Display the full help text")

	if err := fs.Parse(os.Args[1:]); err != nil {
		os.Exit(1)
	}

	if *help {
		DoHelp()
	}

	if *dbname2Arg == "" || *password2Arg == "" || *dbname1Arg == "" || *password1Arg == "" {
		requiredArguments := ""
		if *dbname1Arg == "" {
			requiredArguments += " --dbname1"
		}
		if *password1Arg == "" {
			requiredArguments += " --password1"
		}
		if *dbname2Arg == "" {
			requiredArguments += " --dbname2"
		}
		if *password2Arg == "" {
			requiredArguments += " --password2"
		}
		message := fmt.Sprintf("%s: error: the following arguments are required: %s\n", os.Args[0], requiredArguments)
		DoUsage(message)
	}

	verbose = *verboseArg

	userid1 := *userid1Arg
	if *userid1Arg == "" {
		userid1 = *dbname1Arg
	}
	schema1 := *schema1Arg
	if schema1 == "" {
		schema1 = *userid1Arg
	}

	userid2 := *userid2Arg
	if *userid2Arg == "" {
		userid2 = *dbname2Arg
	}
	schema2 := *schema2Arg
	if schema2 == "" {
		schema2 = *userid2Arg
	}

	firstConnectionString := fmt.Sprintf("HOSTNAME=%s;DATABASE=%s;PORT=%d;UID=%s;PWD=%s", *hostname1Arg, *dbname1Arg, *port1Arg, userid1, *password1Arg)
	secondConnectionString := fmt.Sprintf("HOSTNAME=%s;DATABASE=%s;PORT=%d;UID=%s;PWD=%s", *hostname2Arg, *dbname2Arg, *port2Arg, userid2, *password2Arg)

	type Db *sql.DB
	var firstConn Db
	firstConn = CreateConn(firstConnectionString)
	if firstConn == nil {
		fmt.Printf("Unable to connect successfully to %s!", firstConnectionString)
		os.Exit(1)
	}
	var secondConn Db
	secondConn = CreateConn(secondConnectionString)
	if secondConn == nil {
		fmt.Printf("Unable to connect successfully to %s!", secondConnectionString)
		os.Exit(1)
	}
	err := compareAllEntryModifyTimestamps(firstConn, secondConn, schema1, schema2)
	if err != nil {
		fmt.Println(err)
	}
}

File: IAM/tools/replication/ldap_sdiff.py
import ibm_db
import argparse
import sys
import csv
from datetime import datetime
import logging.config
import logging

logger = logging.getLogger(__name__)


def list_repl_contexts(conn, schema):
    """
    Execute SQL statement to return all DNs and ModifyTimestamps in order of DN
    from a LDAP Server.

    :param conn:
    :param schema:
    :return:
    """
    sql = (
        "select dn, dn_trunc, modify_timestamp - current timezone"  # Return timestamp in UTC format
        " from {}.ldap_entry"
        " order by dn_trunc"
    ).format(schema)
    logger.debug("Executing SQL: {}".format(sql))
    return ibm_db.exec_immediate(conn, sql)


def write_to_csv(writer, dn, dn2, dn_trunc, current_timestamp, modify_timestamp1, modify_timestamp2, status):
    """
    Write to CSV with additional checks to detect truncation issue and/or changes in flight.

    :param writer:
    :param dn:
    :param dn2:
    :param dn_trunc:
    :param current_timestamp:
    :param modify_timestamp1:
    :param modify_timestamp2:
    :param status:
    :return:
    """
    if modify_timestamp1 and modify_timestamp1 > current_timestamp:
        logger.debug("modify_timestamp1 > current_timestamp: {} > {}".format(modify_timestamp1, current_timestamp))
        status += '*'
    if modify_timestamp2 and modify_timestamp2 > current_timestamp:
        logger.debug("modify_timestamp2 > current_timestamp: {} > {}".format(modify_timestamp2, current_timestamp))
        status += '*'
    if dn != dn_trunc:
        logger.debug("DN Truncated: \n  DN - {}, \n  DN_TRUNC - {}".format(dn, dn_trunc))
        status += '#'
    if dn2 and dn != dn2:
        logger.debug("dn != dn2, dn:{} dn2{}".format(dn, dn2))
        status += '#'
    writer.writerow({
        'dn': dn,
        'modifyTimestamp1': modify_timestamp1,
        'modifyTimestamp2': modify_timestamp2,
        'status': status
    })


def print_legend(hostname1, hostname2):
    """
    Print a legend describing what the CSV output looks like.

    :param hostname1:
    :param hostname2:
    :return:
    """
    print("Legend for output:")
    print("  dn - distinguished name of the object that is out of sync or missing.")
    print("  modifyTimestamp1 - dn's object last modified on server: {}.".format(hostname1))
    print("  modifyTimestamp2 - dn's object last modified on server: {}.".format(hostname2))
    print("  status - the following are what the values mean:")
    print("    1: Object detected in both servers but modify timestamps are different indicating mis-match.")
    print("    2: Object found in {} but missing in {}.".format(hostname1, hostname2))
    print("    3: Object found in {} but missing in {}.".format(hostname2, hostname1))
    print("    #: Detecting truncation of DN, further comparison of object recommended.")
    print("    *: Detecting Modify Timestamp more recent than when this script started execution,")
    print("       further comparison of object recommended (replication may fix mis-match).")
    print("")
    print("Note: Timestamps are provided in UTC timezone.")
    print("")


def compare_all_entry_modify_timestamps(hostname1, conn1, schema1, hostname2, conn2, schema2, fout):
    """
    Compare all DNs from one LDAP server with another and print discrepancies.

    Note: dn_trunc is being used for checking discrepancies for highest efficiencies, however
    dn is bring printed in the output csv. There is a small chance that dn_trunc is truncated.

    :param conn1:
    :param conn2:
    :param schema1:
    :param schema2:
    :return:
    """
    current_timestamp = datetime.utcnow()
    count1, count2 = 0, 0
    # Get DNs from first server
    stmt1 = list_repl_contexts(conn1, schema1)
    result1 = ibm_db.fetch_tuple(stmt1)
    if not result1:
        raise Exception("No rows returned for {} server!".format(hostname1))
    # Get DNs from second server
    stmt2 = list_repl_contexts(conn2, schema2)
    result2 = ibm_db.fetch_tuple(stmt2)
    if not result2:
        raise Exception("No rows returned for {} server!".format(hostname2))
    # Compare all DNs from first and second server
    writer = csv.DictWriter(fout, ['dn', 'modifyTimestamp1', 'modifyTimestamp2', 'status'])
    writer.writeheader()
    while result1 and result2:
        dn1, dn_trunc1, modify_timestamp1 = result1[0].strip(), result1[1], result1[2]
        dn2, dn_trunc2, modify_timestamp2 = result2[0].strip(), result2[1], result2[2]
        count1 += 1
        count2 += 1
        if dn_trunc1 == dn_trunc2:
            if modify_timestamp1 != modify_timestamp2:
                # dn2 and dn2 should be same
                write_to_csv(writer, dn1, dn2, dn_trunc1, current_timestamp, modify_timestamp1, modify_timestamp2, '1')
            result1 = ibm_db.fetch_tuple(stmt1)
            result2 = ibm_db.fetch_tuple(stmt2)
        elif dn_trunc1 < dn_trunc2:
            write_to_csv(writer, dn1, None, dn_trunc1, current_timestamp, modify_timestamp1, None, '2')
            result1 = ibm_db.fetch_tuple(stmt1)
        else:
            write_to_csv(writer, dn2, None, dn_trunc2, current_timestamp, None, modify_timestamp2, '3')
            result2 = ibm_db.fetch_tuple(stmt2)
    # Report on extra DNs in first server
    while result1:
        count1 += 1
        dn_trunc1, modify_timestamp1 = result1[0], result1[1]
        write_to_csv(writer, dn1, None, dn_trunc1, current_timestamp, modify_timestamp1, None, '2')
        result1 = ibm_db.fetch_tuple(stmt1)
    # Report on extra DNs in second server
    while result2:
        count2 += 1
        dn_trunc2, modify_timestamp2 = result2[0], result2[1]
        write_to_csv(writer, dn2, None, dn_trunc2, current_timestamp, None, modify_timestamp2, '3')
        result2 = ibm_db.fetch_tuple(stmt2)
    print("Total Entries in {}: {}".format(hostname1, count1))
    print("Total Entries in {}: {}".format(hostname2, count2))


def get_connection(dbname, hostname, port, userid, password):
    conn_str = (
        "DATABASE={};"
        "HOSTNAME={};"
        "PORT={};"
        "PROTOCOL=TCPIP;"
        "UID={};"
        "PWD={};"
    ).format(dbname, hostname, port, userid, password)
    logger.debug("DB2 Connection: {}".format(conn_str))
    return ibm_db.pconnect(conn_str, "", "")


def get_arguments():
    """
    Get the command-line arguments
    """
    aparser = argparse.ArgumentParser(description='Provide DB2 connection details to compare LDAP server contents.')
    aparser.add_argument('--dbname1', help='DB2 Database Name underlying LDAP.', required=True)
    aparser.add_argument('--hostname1', help='Hostname of LDAP server (defaults to localhost).', default='localhost')
    aparser.add_argument('--port1', help='Port# DB2 is listening on (defaults to 50000).', default=50000)
    aparser.add_argument('--schema1', help='DB2 Table name schema (defaults to userid1).', required=False)
    aparser.add_argument('--userid1', help='Userid to connect to DB2 (defaults to dbname1).', required=False)
    aparser.add_argument('--password1', help='Password to connect to DB2.', required=True)
    aparser.add_argument('--dbname2', help='DB2 Database Name underlying LDAP (defaults to dbname1).', required=False)
    aparser.add_argument('--hostname2', help='Hostname of LDAP server (defaults to hostname1).', required=False)
    aparser.add_argument('--port2', help='Port# DB2 is listening on (defaults to port1).', required=False)
    aparser.add_argument('--schema2', help='DB2 Table name schema (defaults to schema1).', required=False)
    aparser.add_argument('--userid2', help='Userid to connect to DB2 (defaults to userid1).', required=False)
    aparser.add_argument('--password2', help='Password to connect to DB2 (defaults to password1).', required=False)
    aparser.add_argument('--loglevel', help='Logging Level (defaults to CRITICAL).', required=False, default='CRITICAL',
                         choices=['DEBUG', 'INFO', 'ERROR', 'CRITICAL'], type=str.upper)
    aparser.add_argument('--output_file', help='Output CSV of differences (defaults to stdout).', required=False)

    try:
        return aparser.parse_args()
    except IOError as msg:
        aparser.error(str(msg))


if __name__ == '__main__':
    try:
        starttime = datetime.utcnow()
        conn1, conn2, fout = None, None, None
        args = get_arguments()
        DEFAULT_LOGGING = {
            'version': 1,
            'disable_existing_loggers': False,
            'formatters': {
                'standard': {
                    'format': '[%(asctime)s] [PID:%(process)d TID:%(thread)d] [%(levelname)s] [%(name)s] [%(funcName)s():%(lineno)s] %(message)s'
                },
            },
            'handlers': {
                'default': {
                    'level': args.loglevel,
                    'formatter': 'standard',
                    'class': 'logging.StreamHandler',
                },
            },
            'loggers': {
                '': {
                    'level': args.loglevel,
                    'handlers': ['default'],
                    'propagate': True
                }
            }
        }
        logging.config.dictConfig(DEFAULT_LOGGING)
        logger.info("Start of Script: {}".format(starttime))
        # Get connection details to DB2 database underlying first LDAP server
        if args.userid1:
            userid1 = args.userid1
        else:
            userid1 = args.dbname1
        if args.schema1:
            schema1 = args.schema1
        else:
            schema1 = userid1
        conn1 = get_connection(args.dbname1, args.hostname1, args.port1, userid1, args.password1)

        # Get connection details to DB2 database underlying second LDAP server
        if args.dbname2:
            dbname2 = args.dbname2
        else:
            dbname2 = args.dbname1
        if args.hostname2:
            hostname2 = args.hostname2
        else:
            hostname2 = args.hostname1
        if args.port2:
            port2 = args.port2
        else:
            port2 = args.port1
        if args.userid2:
            userid2 = args.userid2
        else:
            userid2 = userid1
        if args.password2:
            password2 = args.password2
        else:
            password2 = args.password1
        if args.schema2:
            schema2 = args.schema2
        else:
            schema2 = schema1
        conn2 = get_connection(dbname2, hostname2, port2, userid2, password2)

        if args.output_file:
            fout = open(args.output_file, 'w')
        else:
            fout = sys.stdout
        print_legend(args.hostname1, hostname2)
        # Compare DNs/ModifyTimestamp from first and second LDAP servers
        compare_all_entry_modify_timestamps(args.hostname1, conn1, schema1, hostname2, conn2, schema2, fout)
        endtime = datetime.utcnow()
        logger.info("End of Script: {}".format(endtime))
        print("Script ran for: {}".format(endtime - starttime))
    except Exception as e:
        conn_error = ibm_db.conn_error()
        stmt_error = ibm_db.stmt_error()
        if conn_error != '':
            print("Error Code: {} Msg: {}".format(conn_error, ibm_db.conn_errormsg()))
            if conn1:
                print("Connection issue with server#2 most probably.")
            else:
                print("Connection issue with server#1 most probably.")
        elif stmt_error != '':
            print("Error Code: {} Msg: {}".format(stmt_error, ibm_db.stmt_errormsg()))
        raise e
    finally:
        if fout and fout is not sys.stdout:
            fout.close()
        if conn1:
            ibm_db.close(conn1)
        if conn2:
            ibm_db.close(conn2)
